### PR TITLE
Push data to HK, rather than wait for it to be pulled 

### PR DESCRIPTION
**Pull request by davidjbradshaw**
This PR makes the following changes

 * Push realtime data to HomeBridge, rather than just waiting for it to be pulled
 * Improved logging
    * Power disconnect/reconnect is now a warning
    * Battery level now shows time remaining and voltage
    * warn/error now also have code to stop the same message being shown twice
 * Code refactor
    * Created parse functions, to allow code reuse
    * Upgraded to latest version of esLint/prettier
    * `log.update` -> `log.update.info`
    * Defined consts for bitwise values
    * Defined consts for default values
    * Defined consts for magic numbers
 * Updated README to show default values

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -6,10 +6,7 @@ const { logOnlyError, logMin } = require('./lib/logUpdate');
 class APCAccess {
   constructor(log, config) {
     this.config = config;
-    this.log = logMin(log);
-    if (config.errorLogsOnly === true) {
-      this.log = logOnlyError(log);
-    }
+    this.log = config.errorLogsOnly ? logOnlyError(log) :logMin(log);
```
Comment (cr3ative) [inline index.js:9]: It's lean! It's clean! It's not just two, but three less lines of code!

File: index.js [hunk 2]
```diff
@@ -31,31 +28,36 @@ class APCAccess {
 
     // The following can't be defined on boot, so define them optionally in config
     this.contactSensor = new Service.ContactSensor(config.name || 'APCAccess UPS');
+    this.contactSensor
+      .getCharacteristic(Characteristic.ContactSensorState)
+      .on('get', this.getContactState.bind(this));
```
Comment (cr3ative) [inline index.js:33]: Ah, yoinked up from below? Looks much tidier.

File: index.js [hunk 3]
```diff
@@ -31,31 +28,36 @@ class APCAccess {
...
         Characteristic.Manufacturer,
         config.manufacturer || 'American Power Conversion',
       )
+      // I see MODEL: 'Smart-UPS 750 ', in the data
       .setCharacteristic(Characteristic.Model, config.model || 'APCAccess UPS')
+
+      // At least for my device I'm seeing a S/N being return in the data
+      // I'm also seeing Firmware version, which can also be set as a characteristic in HK
+      // SERIALNO: 'AS1539123101  ',
+      // FIRMWARE: 'UPS 09.3 / ID=18',
```
Comment (cr3ative) [inline index.js:47]: Yeah, these values can be pulled but would require a rewrite (a la feature sniffing) to defer establishing the accessory until after the first value set is read.

File: index.js [hunk 4]
```diff
@@ -31,31 +28,36 @@ class APCAccess {
...
-      .on('get', this.getContactState.bind(this));
     this.batteryService
       .getCharacteristic(Characteristic.BatteryLevel)
-      .on('get', this.getBatteryLevel.bind(this));
-    this.batteryService
+      .on('get', this.getBatteryLevel.bind(this))
       .getCharacteristic(Characteristic.ChargingState)
-      .on('get', this.getChargingState.bind(this));
-    this.batteryService
+      .on('get', this.getChargingState.bind(this))
```
Comment (cr3ative) [inline index.js:56]: Blimey, that's clever.

File: index.js [hunk 5]
```diff
@@ -31,31 +28,36 @@ class APCAccess {
...
       .getCharacteristic(Characteristic.StatusLowBattery)
       .on('get', this.getStatusLowBattery.bind(this));
 
+    if(!this.config.temperatureSensor) return;
```
Comment (cr3ative) [inline index.js:60]: Love an early return.

File: index.js [hunk 6]
```diff
@@ -65,29 +67,34 @@ class APCAccess {
   getServices() {
     // Required by Homebridge; expose services this accessory claims to have
     const services = [this.informationService, this.contactSensor, this.batteryService];
-    if (this.config.temperatureSensor === true) {
+
+    if (this.temperatureService) {
```
Comment (cr3ative) [inline index.js:71]: nice

File: index.js [hunk 7]
```diff
@@ -65,29 +67,34 @@ class APCAccess {
...
     return services;
   }
 
   getLatestJSON() {
     this.client.getStatusJson().then((result) => {
       this.latestJSON = result;
       this.doPolledChecks();
+    })
+    .catch((err) => {
+      this.log.error("Polling device failed:", err);
```
Comment (cr3ative) [inline index.js:84]: Nice!

File: index.js [hunk 8]
```diff
@@ -65,29 +67,34 @@ class APCAccess {
...
     });
   }
 
   getBatteryLevel(callback) {
     // BCHARGE
     let battPctValue = 0;
     const battVal = this.latestJSON.BCHARGE;
-    if (battVal !== undefined) {
+
+    if (battVal) {
```
Comment (cr3ative) [inline index.js:93]: This I'm nervous of, a previous bug report regarding this value's incoming data was a nightmare to get stable.

File: index.js [hunk 9]
```diff
@@ -100,35 +107,44 @@ class APCAccess {
...
   }
 
   getContactState(callback) {
     // STATFLAG
     const value = [this.latestJSON.STATFLAG & 0x08 ? 'CONTACT_DETECTED' : 'CONTACT_NOT_DETECTED'];
+
+    // it would be good to log out start time as a warning here
+    if (value === 'CONTACT_NOT_DETECTED') {
+      this.warn('UPS Active', this.latestJSON.STARTTIME)
+    }
```
Comment (cr3ative) [inline index.js:130]: Ooh sure why not

**Review by cr3ative — COMMENTED** [review]
Some notes back atcha!

File: index.js (hunk 3 above)
Comment (davidjbradshaw) [reply index.js:47]: Might be a simpler way of solving this. I'll have a think

**Comment by davidjbradshaw** [comment]
I have it working again now, hope this isn't too much change.

Next steps:
 * Push battery level changes in real time
 * Populate model and firmware data after first data fetch

**Comment by cr3ative** [comment]
> Trying to workout what this does?
> 
> ```js
> 
> const battVal = this.latestJSON.BCHARGE;
> 
> const battArray = battVal.split('.');
> 
> battPctValue = parseFloat(parseFloat(battArray[0] * -1) * -1);
> 
> ```
> 
> and why it might be different to this:
> 
> ```js
> 
> const percentage = parseInt(this.latestJSON.BCHARGE, 10);
> 
> ```

In previous issues, it was found that this level can (but doesn't always) come through as "90.56%" which breaks parseInt. Another oddity of stability.

**Comment by davidjbradshaw** [comment]
> In previous issues, it was found that this level can (but doesn't always) come through as "90.56%" which breaks parseInt. Another oddity of stability.
```js
parseInt("90.56%", 10) // = 90
```
That doesn't make any sense and if that is the case why did it only break it in one of two places you where calling it?

**Comment by davidjbradshaw** [comment]
If the issue is that you want the percentages, then this would work better than `parseInt()`.
```js
parseFloat("90.56% and some other stuff") // = 90.56
```

**Comment by davidjbradshaw** [comment]
I just read #10 and he never say what the issue is that his weird change fixes

**Comment by cr3ative** [comment]
> I just read #10 and he never says what the issue is that his weird change fixes

Happy to change it, it was a bit odd at the time.

**Comment by davidjbradshaw** [comment]
I've put some guard rails around it
```is
  parseBatteryLevel  = () =>  this.loaded ? parseInt(this.latestJSON.BCHARGE, 10) : 0;
  parseTimeLeft = ( Minutes) => this.loaded ? parseInt(this.latestJSON.TIMELEFT, 10) : 0;
  ```

**Comment by davidjbradshaw** [comment]
battery data is now being pushed to HK in real time and I also spotted a bug, in that the low battery warning was attached to the `ContactSensorService`, rather than the `BatteryService`.

```sh
[25/02/2023, 16:47:23] [APCAccess] Launched child bridge with PID 3895089
[25/02/2023, 16:47:24] Registering accessory 'homebridge-apcaccess.APCAccess'
[25/02/2023, 16:47:24] [APCAccess] Loaded homebridge-apcaccess v0.2.2-rc4 child bridge successfully
[25/02/2023, 16:47:24] Loaded 0 cached accessories from cachedAccessories.0EB7A5DE703D.
[25/02/2023, 16:47:24] [APCAccess] Connected!
[25/02/2023, 16:47:24] Homebridge v1.6.0 (HAP v0.11.0) (APCAccess) is running on port 33875.
[25/02/2023, 16:47:25] [APCAccess] Power: Disconnected
[25/02/2023, 16:47:25] [APCAccess] Battery state: Normal
[25/02/2023, 16:47:25] [APCAccess] Charging state: NOT_CHARGING
[25/02/2023, 16:47:25] [APCAccess] Battery Level: 63% (75 estimated minutes remaining)
[25/02/2023, 16:48:51] [APCAccess] Battery Level: 62% (74 estimated minutes remaining)
[25/02/2023, 16:49:34] [APCAccess] Battery Level: 62% (70 estimated minutes remaining)
```

**Comment by davidjbradshaw** [comment]
I've just updated the top with the changes I've made. This might be enough for one PR.

I don't know what to do with the temp sensor, as I have no way of testing it. I could just be careful, but do you have someone that can test that code before release?

Still to do it adding in the Model, S/N and Firmeare to HK from the data once it is loaded in. My thinking is we just put placeholder data in at start up and then on the first pull of the data, the real information can be a single time push.

Thoughts?

**Comment by cr3ative** [comment]
I think careful pushes are fine without explicit testing. I'll take a look at this on Monday on the big screens - likely I'll merge, test it a little, do some very minor tweaks, then release. 

Thank you for your care for this little plugin!

**Comment by davidjbradshaw** [comment]
Tried to get it to update the accessory information once we have the data, but not seeing it push through to the home app. I’ve asked a question about this on Stack Overflow.

**Comment by davidjbradshaw** [comment]
I think I'm going to leave this hear now, don't really want to play with the temp stuff without being able to test it (the patten from the battery level should mostly be the same, if you want to make it realtime).

It's been useful, I think I have learnt enough about home bridge to write my on plugin from play with this.

This is almost a verified plugin now, I think the only out standing task is to implement the forms config for the web UI.

**Comment by cr3ative** [comment]
Sure thing! I'll consider it RFR now and look as soon as I can. :)

File: index.js [hunk 10]
```diff
@@ -1,51 +1,66 @@
...
 
+const DEFAULT_INTERVAL = 1;
+const DEFAULT_NAME = 'APC UPS';
+const DEFAULT_MANIFACTURER = 'American Power Conversion';
+const DEFAULT_MODEL = 'APCAccess UPS';
+const DEFAULT_PORT = '3551';
+
+const FULLY_CHARGED = 100;
+const SECOND = 1000;
+const UNKOWN = 'unkown';
```
Comment (cr3ative) [inline index.js:12]: nit/typo: `UNKNOWN / unknown`

File: index.js [hunk 11]
```diff
@@ -1,51 +1,66 @@
-let Service;
-let Characteristic;
 const ApcAccess = require('apcaccess');
 const { logOnlyError, logMin } = require('./lib/logUpdate');
 
+const DEFAULT_INTERVAL = 1;
+const DEFAULT_NAME = 'APC UPS';
+const DEFAULT_MANIFACTURER = 'American Power Conversion';
```
Comment (cr3ative) [inline index.js:6]: nit/typo: `DEFAULT_MANUFACTURER`

**Review by cr3ative — APPROVED** [review]
Couple of typos, but if you've tested this, I'm happy to test, merge, and publish as a semver major (1.0.0).

File: index.js (hunk 10 above)
Comment (davidjbradshaw) [reply index.js:12]: Dyslexia stricts again. I Will fix in a bit

**Comment by davidjbradshaw** [comment]
Fixed typos, Tested by manually copying files to my rPi and restarting plugin. Then killed the power for a movement. You can see the effect of buying cheap replacement batteries :)

```log
[01/03/2023, 10:29:30] [APCAccess] Restarting Process...
[01/03/2023, 10:29:31] [APCAccess] Launched child bridge with PID 491
[01/03/2023, 10:29:31] Registering accessory 'homebridge-apcaccess.APCAccess'
[01/03/2023, 10:29:31] [APCAccess] Loaded homebridge-apcaccess v0.2.2-rc4 child bridge successfully
[01/03/2023, 10:29:31] Loaded 0 cached accessories from cachedAccessories.0EB7A5DE703D.
[01/03/2023, 10:29:31] [APCAccess] Connected!
[01/03/2023, 10:29:31] Homebridge v1.6.0 (HAP v0.11.0) (APCAccess) is running on port 33875.
[01/03/2023, 10:29:32] [APCAccess] Power: Connected
[01/03/2023, 10:29:32] [APCAccess] Battery state: Normal
[01/03/2023, 10:29:32] [APCAccess] Charging state: CHARGING
[01/03/2023, 10:29:32] [APCAccess] Model: Smart-UPS 750
[01/03/2023, 10:29:32] [APCAccess] Name: home-bridge-1
[01/03/2023, 10:29:32] [APCAccess] SerialNumber: AS1539123101
[01/03/2023, 10:29:32] [APCAccess] SoftwareRevision: 3.14.14 (31 May 2016) debian
[01/03/2023, 10:29:32] [APCAccess] FirmwareRevision: UPS 09.3 / ID=18
[01/03/2023, 10:29:33] [APCAccess] Charging state: NOT_CHARGING
[01/03/2023, 10:29:33] [APCAccess] Battery Level: 100%  27.0 Volts (120 estimated minutes remaining)
[01/03/2023, 10:30:41] [APCAccess] Power: Disconnected
[01/03/2023, 10:30:42] [APCAccess] Battery Level: 97%  27.0 Volts (118 estimated minutes remaining)
[01/03/2023, 10:30:44] [APCAccess] Battery Level: 91%  27.0 Volts (109 estimated minutes remaining)
[01/03/2023, 10:30:46] [APCAccess] Battery Level: 89%  25.4 Volts (106 estimated minutes remaining)
[01/03/2023, 10:30:49] [APCAccess] Battery Level: 84%  25.4 Volts (101 estimated minutes remaining)
[01/03/2023, 10:30:51] [APCAccess] Battery Level: 80%  25.0 Volts (96 estimated minutes remaining)
[01/03/2023, 10:30:53] [APCAccess] Battery Level: 76%  25.0 Volts (91 estimated minutes remaining)
[01/03/2023, 10:30:55] [APCAccess] Battery Level: 73%  25.0 Volts (87 estimated minutes remaining)
[01/03/2023, 10:31:05] [APCAccess] Power: Connected
[01/03/2023, 10:31:05] [APCAccess] Charging state: CHARGING
```

**Comment by cr3ative** [comment]
> Fixed typos, Tested by manually copying files to my rPi and restarting plugin. Then killed the power for a movement. You can see the effect of buying cheap replacement batteries :)
> 
> ```
> [01/03/2023, 10:29:30] [APCAccess] Restarting Process...
> [01/03/2023, 10:29:31] [APCAccess] Launched child bridge with PID 491
> [01/03/2023, 10:29:31] Registering accessory 'homebridge-apcaccess.APCAccess'
> [01/03/2023, 10:29:31] [APCAccess] Loaded homebridge-apcaccess v0.2.2-rc4 child bridge successfully
> [01/03/2023, 10:29:31] Loaded 0 cached accessories from cachedAccessories.0EB7A5DE703D.
> [01/03/2023, 10:29:31] [APCAccess] Connected!
> [01/03/2023, 10:29:31] Homebridge v1.6.0 (HAP v0.11.0) (APCAccess) is running on port 33875.
> [01/03/2023, 10:29:32] [APCAccess] Power: Connected
> [01/03/2023, 10:29:32] [APCAccess] Battery state: Normal
> [01/03/2023, 10:29:32] [APCAccess] Charging state: CHARGING
> [01/03/2023, 10:29:32] [APCAccess] Model: Smart-UPS 750
> [01/03/2023, 10:29:32] [APCAccess] Name: home-bridge-1
> [01/03/2023, 10:29:32] [APCAccess] SerialNumber: AS1539123101
> [01/03/2023, 10:29:32] [APCAccess] SoftwareRevision: 3.14.14 (31 May 2016) debian
> [01/03/2023, 10:29:32] [APCAccess] FirmwareRevision: UPS 09.3 / ID=18
> [01/03/2023, 10:29:33] [APCAccess] Charging state: NOT_CHARGING
> [01/03/2023, 10:29:33] [APCAccess] Battery Level: 100%  27.0 Volts (120 estimated minutes remaining)
> [01/03/2023, 10:30:41] [APCAccess] Power: Disconnected
> [01/03/2023, 10:30:42] [APCAccess] Battery Level: 97%  27.0 Volts (118 estimated minutes remaining)
> [01/03/2023, 10:30:44] [APCAccess] Battery Level: 91%  27.0 Volts (109 estimated minutes remaining)
> [01/03/2023, 10:30:46] [APCAccess] Battery Level: 89%  25.4 Volts (106 estimated minutes remaining)
> [01/03/2023, 10:30:49] [APCAccess] Battery Level: 84%  25.4 Volts (101 estimated minutes remaining)
> [01/03/2023, 10:30:51] [APCAccess] Battery Level: 80%  25.0 Volts (96 estimated minutes remaining)
> [01/03/2023, 10:30:53] [APCAccess] Battery Level: 76%  25.0 Volts (91 estimated minutes remaining)
> [01/03/2023, 10:30:55] [APCAccess] Battery Level: 73%  25.0 Volts (87 estimated minutes remaining)
> [01/03/2023, 10:31:05] [APCAccess] Power: Connected
> [01/03/2023, 10:31:05] [APCAccess] Charging state: CHARGING
> ```

Outstanding! This is amazing work. Going to merge - feel free to tweak further in a new PR, let's land this.

**Comment by cr3ative** [comment]
Published: `+ homebridge-apcaccess@1.0.0`

**Comment by davidjbradshaw** [comment]
Just installed it ok, but you need to update the CHANGELOG when you have a moment.

**Comment by cr3ative** [comment]
> Just installed it ok, but you need to update the CHANGELOG when you have a moment.

Perfect, that's done.

Sent something your way. Get them done! https://secure.ups-trader.co.uk/rebuild-kits/57-rbc48-kit.html?search_query=rbc48&results=5

**Comment by davidjbradshaw** [comment]
Oh wow thank you, I wasn't expecting that.

**Pull request merged**